lib/Dialect/Seq/Transforms/LowerSeqShiftReg.cpp

```diff
@@ -0,0 +1,71 @@
+//===- LowerSeqShiftReg.cpp - seq.shiftreg lowering -----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "circt/Dialect/Seq/SeqPasses.h"
+#include "circt/Support/BackedgeBuilder.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace seq;
+
+namespace {
+
+struct ShiftRegLowering : public OpConversionPattern<seq::ShiftRegOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(seq::ShiftRegOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    Value in = adaptor.getInput();
+    Type dataType = in.getType();
+    for (size_t i = 0; i < op.getSize(); ++i)
+      in = rewriter.create<seq::CompRegClockEnabledOp>(
```

Comment:
previous question of why not lower to`reg g = {g[1:0], g[max:1]`

Reply:
This is not intended to be a canonical lowering, especially not when considering the scope of possible shift register lowerings (stacked registers, single register with shift operation, library instantiation, ...).

This is one possible lowering (the stacked register model facilitates some forms of analysis, but is on the verbose side) mainly motivated by having _some_ lowering available, so we don't have dangling ops in the core dialect. I fully expect that users will bring their own lowerings down the road when it is found that whatever is in CIRCT currently doesn't suffice their needs.

Reply:
I agree there has to be one implementation, but it should be a reasonable default.  My question is "what should the default lowering be?"  Why a bunch of registers vs some bit shuffling?